stablehlo/transforms/StablehloLegalizeQuantizedOpUsingQDQ.cpp

```diff
@@ -0,0 +1,192 @@
+/* Copyright 2024 The StableHLO Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Quant/QuantTypes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"  // Include for TypeConverter
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "stablehlo/dialect/StablehloOps.h"
+#include "stablehlo/transforms/Passes.h"
+
+namespace mlir {
+namespace stablehlo {
+
+#define GEN_PASS_DEF_STABLEHLOLEGALIZEQUANTIZEDOPUSINGQDQPASS
+#include "stablehlo/transforms/Passes.h.inc"
+
+namespace {
+
+bool isAnyQuantizedTypes(TypeRange types) {
+  return llvm::any_of(types, [](Type type) {
+    return isa<quant::QuantizedType>(getElementTypeOrSelf(type));
+  });
+}
+
+#define DEFINE_QUANT_REWRITE_TO_QDQ_PATTERN(OpName)                            \
```

Comment:
Let's write this as a template class then either use a very short macro to just instantiate the macro, or no macro at all if it's short.

Reply:
Take a look at `populateHloToStablehloPatterns`, perhaps we can do a templated populate method like this which takes in a list of `populate<AddOp, DivOp,...>()`

Reply:
done.